src/pushsource/_impl/model/base.py

```diff
@@ -78,3 +96,67 @@ class PushItem(object):
     Generally a short key ID such as "F21541EB" is used, though the library
     doesn't enforce this.
     """
+
+    def with_checksums(self):
+        """Return a copy of this push item with checksums present.
+
+        Many pushsource backends will produce push items with empty checksum attributes,
+        since checksum calculation may be expensive and may not be required in all cases.
+
+        The rule of thumb is that a pushsource will only provide checksums if they were
+        available in metadata, so that reading the push item's file was unnecessary.
+        For example:
+
+        - with ``errata`` push source, checksums are generally available by default, because
+          Errata Tool itself provides checksums on the files contained in an advisory.
+        - with ``staged`` push source, checksums are unavailable by default, because the
+          sums can only be calculated by reading file content from the staging area.
+
+        Where checksums are needed, this utility method may be used to ensure they are present
+        regardless of the push source used.
+
+        This method entails reading the entire content of the file referenced
+        by this push item, and may be:
+
+        - slow: may need to read a large file from a remote system, e.g. from an NFS volume.
+        - error-prone: reads from a remote system might fail.
+
+        As such, when dealing with a large number of push items, you may want to consider
+        using multiple threads to parallelize calls to ``with_checksums``, and retrying
+        failing operations.
+
+        If checksums are already present or if this item does not reference a file, this
+        method is a no-op and returns the current push item, unmodified.
+
+        Returns:
+            :class:`~pushsource.PushItem`
+                A copy of this item, guaranteed either to have non-empty :meth:`md5sum` and
+                :meth:`sha256sum` attributes, or an empty :meth:`src` attribute (denoting that
+                the item does not reference a file).
+
+        .. versionadded:: 1.2.0
+        """
+        if not self.src:
+            return self
+
+        if self.md5sum and self.sha256sum:
+            return self
+
+        md5_hasher = hashlib.md5()
+        sha256_hasher = hashlib.sha256()
+
+        LOG.debug("Start read: %s", self.src)
+
+        with open(self.src, "rb") as src_file:
+            while True:
+                chunk = src_file.read(CHUNKSIZE)
+                if not chunk:
+                    break
+                md5_hasher.update(chunk)
```

Comment:
Probably dumb question, but is it possible that only one of these checksums will already be present? If it could, then some computation can be saved.

Reply:
Yeah, it is possible.  I've assumed (without testing) that the time reading the file here far outweighs the time calculating the checksum, meaning it's not worth trying to skip individual checksum types if any type needs to be calculated.  I'll test that assumption and report back here.

Reply:
I tested it and I was wrong, based on https://gist.github.com/rohanpm/2c1662e39fb138869cd4901cc9b2f9d4, there could be a substantial gain from avoiding the recalculation of checksum types we already have.  So this should be revised.